pkg/health/server/prober.go

```diff
@@ -234,34 +242,105 @@ func (p *prober) setNodes(added nodeMap, removed nodeMap) {
+	// ping each node once, deduplicate nodes into map of nodeName -> []IP.
+	nodes := make(map[string][]*net.IPAddr)
+	for _, node := range p.nodes {
+		if nodes[node.Name] != nil {
+			// Already handled this node.
+			continue
+		}
+		nodes[node.Name] = []*net.IPAddr{}
+		for elem, primary := range node.Addresses() {
+			if _, addr := resolveIP(&node, elem, primary); addr != nil {
+				nodes[node.Name] = append(nodes[node.Name], addr)
+			}
 		}
 	}
+
+	return nodes
 }
 
-func (p *prober) clearIcmpStatus() {
-	p.Lock()
-	defer p.Unlock()
+func (p *prober) icmpPing(node string, ip string, ctx context.Context, resChan chan<- connectivityResult, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	result := &models.ConnectivityStatus{}
+	if err := p.icmpPingerRateLimiter.Wait(ctx); err != nil {
+		result.Status = err.Error()
+		resChan <- connectivityResult{ip: ip, status: result}
+		return
+	}
+
+	// Only add fields to the scoped logger if debug is enabled, to save on resources.
+	// This can be done since all logs in this function are debug-level only.
+	scopedLog := log
+	debugLogsEnabled := logging.CanLogAt(scopedLog.Logger, logrus.DebugLevel)
+	if debugLogsEnabled {
+		scopedLog = log.WithFields(logrus.Fields{
+			logfields.NodeName: node,
+			logfields.IPAddr:   ip,
+		})
+	}
+
+	scopedLog.Debug("Pinging host")
+	pinger, err := probing.NewPinger(ip)
+	if err != nil {
+		if debugLogsEnabled {
+			scopedLog.WithError(err).Debug("Failed to create pinger")
+		}
+		result.Status = err.Error()
+		resChan <- connectivityResult{ip: ip, status: result}
+		return
+	}
+
+	pinger.Timeout = p.server.ProbeDeadline
+	pinger.Count = 1
+	pinger.OnFinish = func(stats *probing.Statistics) {
+		p.Lock()
+		defer p.Unlock()
 
-	for _, status := range p.results {
-		status.Icmp = nil
+		if stats.PacketsRecv > 0 && len(stats.Rtts) > 0 {
+			if debugLogsEnabled {
+				scopedLog.WithField("rtt", stats.Rtts[0].Nanoseconds()).Debug("Ping successful")
+			}
+			result.Latency = stats.Rtts[0].Nanoseconds()
+			result.Status = ""
+		} else {
+			scopedLog.Debug("Ping failed")
+			result.Status = "Connection timed out"
+		}
 	}
+	pinger.SetPrivileged(true)
+	err = pinger.Run()
+	if err != nil {
+		scopedLog.Debugf("Failed to run pinger for IP %s: %v", ip, err)
+		result.Status = err.Error()
+	}
+	resChan <- connectivityResult{ip: ip, status: result}
 }
 
-const httpPathDescription = "Via L3"
+func Per(nodes int, duration time.Duration) rate.Limit {
+	return rate.Every(duration / time.Duration(nodes))
+}
+
+func (p *prober) httpProbe(node string, ip string, ctx context.Context, resChan chan<- connectivityResult, wg *sync.WaitGroup) {
+	defer wg.Done()
 
-func (p *prober) httpProbe(node string, ip string) *models.ConnectivityStatus {
 	result := &models.ConnectivityStatus{}
+	if err := p.httpProbeRateLimiter.Wait(ctx); err != nil {
+		result.Status = err.Error()
+		resChan <- connectivityResult{ip: ip, status: result}
+		return
+	}
```

Comment:
Similar to `icmpPing`, if we move out `rateLimiter` and pass `p.server.Config.HTTPPathPort` through arguments, we could change it from method to function.